shinylive/_url.py

```diff
@@ -23,179 +25,598 @@ class FileContentJson(TypedDict):
+            other_path = str(Path(file).absolute().relative_to(root_dir))
+            if other_path in file_names:
+                index = file_names.index(other_path)
 
-        for file in files:
-            if Path(file).is_dir():
-                file_list.extend(listdir_recursive(file))
-            else:
-                file_list.append(file)
+        if index is None:
+            raise ValueError(f"File '{file}' not found in app bundle.")
 
-        file_bundle = file_bundle + [
-            read_file(file, root_dir) for file in file_list if Path(file) != app_path
-        ]
+        return self
 
-    return file_bundle
+    def __add__(self, other: str | Path) -> ShinyliveApp:
+        other = Path(other)
+        new: ShinyliveApp = copy.deepcopy(self)
+        if other.is_dir():
+            new.add_dir(other)
+        else:
+            new.add_file(other)
+        return new
 
+    def __sub__(self, other: str | Path) -> ShinyliveApp:
+        new: ShinyliveApp = copy.deepcopy(self)
+        new.remove_file(other)
+        return new
 
-def detect_app_language(app: str | Path) -> Literal["py", "r"]:
-    err_not_detected = """
-    Could not automatically detect the language of the app. Please specify `language`."""
 
-    if isinstance(app, str) and "\n" in app:
-        if re.search(r"^(import|from) shiny", app, re.MULTILINE):
-            return "py"
-        elif re.search(r"^library\(shiny\)", app, re.MULTILINE):
-            return "r"
-        else:
-            raise ValueError(err_not_detected)
+def url_encode(
+    app: str | Path,
+    files: Optional[str | Path | Sequence[str | Path]] = None,
+    language: Optional[Literal["py", "r"]] = None,
+    mode: Literal["editor", "app"] = "editor",
+    header: bool = True,
+    host: str = "https://shinylive.io",
+) -> str:
+    """
+    Generate a URL for a [Shinylive application](https://shinylive.io).
 
-    app = Path(app)
+    Parameters
+    ----------
+    app
+        The main app file of the Shinylive application. This file should be a Python
+        `app.py` or an R `app.R`, `ui.R`, or `server.R` file. This file will be renamed
+         `app.py` or `app.R` for shinylive, unless it's named `ui.R` or `server.R`.
+    files
+        File(s) or directory path(s) to include in the application. On shinylive, these
+        files will be stored relative to the main `app` file.
+    mode
+        The mode of the application, either "editor" or "app". Defaults to "editor".
+    language
+        The language of the application, or None to autodetect the language. Defaults to
+        None.
+    header
+        Whether to include a header bar in the UI. This is used only if ``mode`` is
+        "app". Defaults to True.
 
-    if app.suffix.lower() == ".py":
-        return "py"
-    elif app.suffix.lower() == ".r":
-        return "r"
+
+    Returns
+    -------
+        A ShinyliveIoApp object. Use the `.url()` method to retrieve the Shinylive URL.
```

Comment:
There are a few remaining instances of `ShinyliveIoApp`.

```suggestion
        A ShinyliveApp object. Use the `.url()` method to retrieve the Shinylive URL.
```